KeyError during stream processing

@ligser writes on Gitter:


```
Hello all! I have some troubles with spacy-nightly==2.0.0rc1 (a18 has same behavior) with en_core_web_lg model. When I run nlp.pipe with a generator of a texts I get the exception: KeyError: 4405041669077156115..
Exception raised after amount of texts (average 10000).
Stacktrace looks like that:
    nlp.pipe((c.content_text for c in texts), batch_size=24, n_threads=8)
  File "doc.pyx", line 375, in spacy.tokens.doc.Doc.text.__get__
  File "doc.pyx", line 232, in __iter__ 
  File "token.pyx", line 178, in spacy.tokens.token.Token.text_with_ws.__get__
  File "strings.pyx", line 116, in spacy.strings.StringStore.__getitem__
KeyError: 4405041669077156115
That looks like a bug with a StringStore cleanup or something related (maybe shared string-store that clean-up by one of threads?).
My code just get a texts from mysql split it to texts and ids and do: for id, doc in zip(ids_gen,nlp.pipe(docs_gen, ...)).
```

I think this is likely due to the solution added in spaCy 2 to address the streaming data memory growth.

I made some experiments and reproduce similar error as _test case_. I very new in python, and code looks ugly and work slow, but it cause error that looks same (but have different stacktrace):

```
# coding: utf8
from __future__ import unicode_literals

import random
import string

from ...lang.en import English


def test_issue1506():
    nlp = English()

    def random_string_generator(string_length, limit):
        for _ in range(limit):
            yield ''.join(
                random.choice(string.digits + string.ascii_letters + '. ') for _ in range(string_length))

    for i, d in zip(
        (i for i in range(20007)),
        nlp.pipe(random_string_generator(600, 20007))
    ):
        str(d.text)
```

## Info about spaCy

* **spaCy version:** 2.0.2.dev0
* **Platform:** Darwin-17.2.0-x86_64-i386-64bit
* **Python version:** 3.6.2

Stacktrace:
```
spacy/language.py:554: in pipe
    for doc in docs:
spacy/language.py:534: in <genexpr>
    docs = (self.make_doc(text) for text in texts)
spacy/language.py:357: in make_doc
    return self.tokenizer(text)
tokenizer.pyx:106: in spacy.tokenizer.Tokenizer.__call__
    ???
tokenizer.pyx:156: in spacy.tokenizer.Tokenizer._tokenize
    ???
tokenizer.pyx:235: in spacy.tokenizer.Tokenizer._attach_tokens
    ???
doc.pyx:547: in spacy.tokens.doc.Doc.push_back
    ???
morphology.pyx:81: in spacy.morphology.Morphology.assign_untagged
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   KeyError: 10868232842057966403

strings.pyx:116: KeyError
```

--- a/.github/contributors/ligser.md
+++ b/.github/contributors/ligser.md
@@ -1,0 +1,106 @@
+# spaCy contributor agreement
+
+This spaCy Contributor Agreement (**"SCA"**) is based on the
+[Oracle Contributor Agreement](http://www.oracle.com/technetwork/oca-405177.pdf).
+The SCA applies to any contribution that you make to any product or project
+managed by us (the **"project"**), and sets out the intellectual property rights
+you grant to us in the contributed materials. The term **"us"** shall mean
+[ExplosionAI UG (haftungsbeschränkt)](https://explosion.ai/legal). The term
+**"you"** shall mean the person or entity identified below.
+
+If you agree to be bound by these terms, fill in the information requested
+below and include the filled-in version with your first pull request, under the
+folder [`.github/contributors/`](/.github/contributors/). The name of the file
+should be your GitHub username, with the extension `.md`. For example, the user
+example_user would create the file `.github/contributors/example_user.md`.
+
+Read this agreement carefully before signing. These terms and conditions
+constitute a binding legal agreement.
+
+## Contributor Agreement
+
+1. The term "contribution" or "contributed materials" means any source code,
+object code, patch, tool, sample, graphic, specification, manual,
+documentation, or any other material posted or submitted by you to the project.
+
+2. With respect to any worldwide copyrights, or copyright applications and
+registrations, in your contribution:
+
+    * you hereby assign to us joint ownership, and to the extent that such
+    assignment is or becomes invalid, ineffective or unenforceable, you hereby
+    grant to us a perpetual, irrevocable, non-exclusive, worldwide, no-charge,
+    royalty-free, unrestricted license to exercise all rights under those
+    copyrights. This includes, at our option, the right to sublicense these same
+    rights to third parties through multiple levels of sublicensees or other
+    licensing arrangements;
+
+    * you agree that each of us can do all things in relation to your
+    contribution as if each of us were the sole owners, and if one of us makes
+    a derivative work of your contribution, the one who makes the derivative
+    work (or has it made will be the sole owner of that derivative work;
+
+    * you agree that you will not assert any moral rights in your contribution
+    against us, our licensees or transferees;
+
+    * you agree that we may register a copyright in your contribution and
+    exercise all ownership rights associated with it; and
+
+    * you agree that neither of us has any duty to consult with, obtain the
+    consent of, pay or render an accounting to the other for any use or
+    distribution of your contribution.
+
+3. With respect to any patents you own, or that you can license without payment
+to any third party, you hereby grant to us a perpetual, irrevocable,
+non-exclusive, worldwide, no-charge, royalty-free license to:
+
+    * make, have made, use, sell, offer to sell, import, and otherwise transfer
+    your contribution in whole or in part, alone or in combination with or
+    included in any product, work or materials arising out of the project to
+    which your contribution was submitted, and
+
+    * at our option, to sublicense these same rights to third parties through
+    multiple levels of sublicensees or other licensing arrangements.
+
+4. Except as set out above, you keep all right, title, and interest in your
+contribution. The rights that you grant to us under these terms are effective
+on the date you first submitted a contribution to us, even if your submission
+took place before the date you sign these terms.
+
+5. You covenant, represent, warrant and agree that:
+
+    * Each contribution that you submit is and shall be an original work of
+    authorship and you can legally grant the rights set out in this SCA;
+
+    * to the best of your knowledge, each contribution will not violate any
+    third party's copyrights, trademarks, patents, or other intellectual
+    property rights; and
+
+    * each contribution shall be in compliance with U.S. export control laws and
+    other applicable export and import laws. You agree to notify us if you
+    become aware of any circumstance which would make any of the foregoing
+    representations inaccurate in any respect. We may publicly disclose your
+    participation in the project, including the fact that you have signed the SCA.
+
+6. This SCA is governed by the laws of the State of California and applicable
+U.S. Federal law. Any choice of law rules will not apply.
+
+7. Please place an “x” on one of the applicable statement below. Please do NOT
+mark both statements:
+
+    * [x] I am signing on behalf of myself as an individual and no other person
+    or entity, including my employer, has or will have rights with respect to my
+    contributions.
+
+    * [ ] I am signing on behalf of my employer or a legal entity and I have the
+    actual authority to contractually bind that entity.
+
+## Contributor Details
+
+| Field                          | Entry                              |
+|------------------------------- | --------------------               |
+| Name                           | Roman Domrachev                    |
+| Company name (if applicable)   |                                    |
+| Title or role (if applicable)  |                                    |
+| Date                           | 11 November 2017                   |
+| GitHub username                | ligser                             |
+| Website (optional)             |                                    |

--- a/spacy/language.py
+++ b/spacy/language.py
@@ -12,7 +12,6 @@ from copy import copy
 from thinc.neural import Model
 from thinc.neural.optimizers import Adam
 
-from .strings import StringStore
 from .tokenizer import Tokenizer
 from .vocab import Vocab
 from .lemmatizer import Lemmatizer

--- a/spacy/language.py
+++ b/spacy/language.py
@@ -12,6 +12,7 @@ from copy import copy
 from thinc.neural import Model
 from thinc.neural.optimizers import Adam
 
+from .strings import StringStore
 from .tokenizer import Tokenizer
 from .vocab import Vocab
 from .lemmatizer import Lemmatizer
@@ -547,27 +548,20 @@ class Language(object):
         # in the string store.
         recent_refs = weakref.WeakSet()
         old_refs = weakref.WeakSet()
-        original_strings_data = self.vocab.strings.to_bytes()
-        StringStore = self.vocab.strings.__class__
-        recent_strings = StringStore().from_bytes(original_strings_data)
+        # If there is anything that we have inside — after iterations we should
+        # carefully get it back.
+        original_strings_data = list(self.vocab.strings)
         nr_seen = 0
         for doc in docs:
             yield doc
-            for word in doc:
-                recent_strings.add(word.text)
             recent_refs.add(doc)
             if nr_seen < 10000:
                 old_refs.add(doc)
                 nr_seen += 1
             elif len(old_refs) == 0:
-                # All the docs in the 'old' set have expired, so the only
-                # difference between the backup strings and the current
-                # string-store should be obsolete. We therefore swap out the
-                # old strings data.
-                old_refs, recent_refs = recent_refs, old_refs
-                self.vocab.strings._reset_and_load(recent_strings)
-                recent_strings = StringStore().from_bytes(original_strings_data)
+                self.vocab.strings._cleanup_stale_strings()
                 nr_seen = 0
+        self.vocab.strings._reset_and_load(original_strings_data)
 
     def to_disk(self, path, disable=tuple()):
         """Save the current state to a directory.  If a model is loaded, this

--- a/spacy/language.py
+++ b/spacy/language.py
@@ -560,7 +560,11 @@ class Language(object):
             elif len(old_refs) == 0:
                 self.vocab.strings._cleanup_stale_strings()
                 nr_seen = 0
-        self.vocab.strings._reset_and_load(original_strings_data)
+        # Last batch can be not garbage collected and we cannot know it — last
+        # doc still here. Not erase that strings — just extend with original
+        # content
+        for string in original_strings_data:
+            self.vocab.strings.add(string)
 
     def to_disk(self, path, disable=tuple()):
         """Save the current state to a directory.  If a model is loaded, this

--- a/spacy/language.py
+++ b/spacy/language.py
@@ -560,9 +560,9 @@ class Language(object):
             elif len(old_refs) == 0:
                 self.vocab.strings._cleanup_stale_strings()
                 nr_seen = 0
-        # Last batch can be not garbage collected and we cannot know it — last
-        # doc still here. Not erase that strings — just extend with original
-        # content
+        # We can't know which strings from the last batch have really expired.
+        # So we don't erase the strings — we just extend with the original
+        # content.
         for string in original_strings_data:
             self.vocab.strings.add(string)
 

--- a/spacy/strings.pxd
+++ b/spacy/strings.pxd
@@ -1,5 +1,6 @@
 from libc.stdint cimport int64_t
 from libcpp.vector cimport vector
+from libcpp.set cimport set
 
 from cymem.cymem cimport Pool
 from preshed.maps cimport PreshMap
@@ -23,6 +24,7 @@ cdef class StringStore:
     cdef Pool mem
 
     cdef vector[hash_t] keys
+    cdef set[hash_t] hits
     cdef public PreshMap _map
 
     cdef const Utf8Str* intern_unicode(self, unicode py_string)

--- a/spacy/strings.pyx
+++ b/spacy/strings.pyx
@@ -4,6 +4,7 @@ from __future__ import unicode_literals, absolute_import
 
 cimport cython
 from libc.string cimport memcpy
+from libcpp.set cimport set
 from libc.stdint cimport uint32_t
 from murmurhash.mrmr cimport hash64, hash32
 import ujson
@@ -111,6 +112,7 @@ cdef class StringStore:
             return SYMBOLS_BY_INT[string_or_id]
         else:
             key = string_or_id
+            self.hits.insert(key)
             utf8str = <Utf8Str*>self._map.get(key)
             if utf8str is NULL:
                 raise KeyError(string_or_id)
@@ -168,6 +170,7 @@ cdef class StringStore:
         if key < len(SYMBOLS_BY_INT):
             return True
         else:
+            self.hits.insert(key)
             return self._map.get(key) is not NULL
 
     def __iter__(self):
@@ -179,6 +182,7 @@ cdef class StringStore:
         cdef hash_t key
         for i in range(self.keys.size()):
             key = self.keys[i]
+            self.hits.insert(key)
             utf8str = <Utf8Str*>self._map.get(key)
             yield decode_Utf8Str(utf8str)
         # TODO: Iterate OOV here?
@@ -241,8 +245,23 @@ cdef class StringStore:
         self.mem = Pool()
         self._map = PreshMap()
         self.keys.clear()
+        self.hits.clear()
         for string in strings:
             self.add(string)
+
+    def _cleanup_stale_strings(self):
+        if self.hits.size() == 0:
+            # If no any hits — just skip cleanup
+            return
+
+        cdef vector[hash_t] tmp
+        for i in range(self.keys.size()):
+            key = self.keys[i]
+            if self.hits.count(key) != 0:
+                tmp.push_back(key)
+
+        self.keys.swap(tmp)
+        self.hits.clear()
 
     cdef const Utf8Str* intern_unicode(self, unicode py_string):
         # 0 means missing, but we don't bother offsetting the index.
@@ -259,5 +278,6 @@ cdef class StringStore:
             return value
         value = _allocate(self.mem, <unsigned char*>utf8_string, length)
         self._map.set(key, value)
+        self.hits.insert(key)
         self.keys.push_back(key)
         return value

--- a/spacy/strings.pyx
+++ b/spacy/strings.pyx
@@ -251,7 +251,7 @@ cdef class StringStore:
 
     def _cleanup_stale_strings(self):
         if self.hits.size() == 0:
-            # If no any hits — just skip cleanup
+            # If we don't have any hits, just skip cleanup
             return
 
         cdef vector[hash_t] tmp

--- a/spacy/tests/regression/test_issue1506.py
+++ b/spacy/tests/regression/test_issue1506.py
@@ -1,0 +1,28 @@
+# coding: utf8
+from __future__ import unicode_literals
+
+import random
+import string
+
+import itertools
+from compat import izip
+
+from ...lang.en import English
+
+
+def test_issue1506():
+    nlp = English()
+
+    def string_generator():
+        for (_, t) in izip(range(10001), itertools.repeat("It's sentence produced by that bug.")):
+            yield t
+
+        for (_, t) in izip(range(10001), itertools.repeat("I erase lemmas.")):
+            yield t
+
+        for (_, t) in izip(range(10001), itertools.repeat("It's sentence produced by that bug.")):
+            yield t
+
+    for d in nlp.pipe(string_generator()):
+        for t in d:
+            str(t.lemma_)

--- a/spacy/tests/regression/test_issue1506.py
+++ b/spacy/tests/regression/test_issue1506.py
@@ -1,12 +1,6 @@
 # coding: utf8
 from __future__ import unicode_literals
 
-import random
-import string
-
-import itertools
-from compat import izip
-
 from ...lang.en import English
 
 
@@ -14,14 +8,14 @@ def test_issue1506():
     nlp = English()
 
     def string_generator():
-        for (_, t) in izip(range(10001), itertools.repeat("It's sentence produced by that bug.")):
-            yield t
+        for _ in range(10001):
+            yield "It's sentence produced by that bug."
 
-        for (_, t) in izip(range(10001), itertools.repeat("I erase lemmas.")):
-            yield t
+        for _ in range(10001):
+            yield "I erase lemmas."
 
-        for (_, t) in izip(range(10001), itertools.repeat("It's sentence produced by that bug.")):
-            yield t
+        for _ in range(10001):
+            yield "It's sentence produced by that bug."
 
     for d in nlp.pipe(string_generator()):
         for t in d:

